### ISSUE TITLE
bigtable: potential deadlock on ApplyBulk?

### ISSUE THREAD

**Issue by tomwilkie** [GoogleCloudPlatform/google-cloud-go#797]
I've got a goroutine stuck in ApplyBulk for >28hrs:

```
goroutine 163 [select, 1728 minutes]:
github.com/weaveworks/cortex/vendor/google.golang.org/grpc/transport.(*recvBufferReader).Read(0xc425c4d580, 0xc422910770, 0x5, 0x5, 0x0, 0x0, 0x0)
	/go/src/github.com/weaveworks/cortex/vendor/google.golang.org/grpc/transport/transport.go:144 +0x556
github.com/weaveworks/cortex/vendor/google.golang.org/grpc/transport.(*Stream).Read(0xc42d2d90e0, 0xc422910770, 0x5, 0x5, 0xc430257458, 0x9, 0x1800)
	/go/src/github.com/weaveworks/cortex/vendor/google.golang.org/grpc/transport/transport.go:332 +0x5c
io.ReadAtLeast(0x154b800, 0xc42d2d90e0, 0xc422910770, 0x5, 0x5, 0x5, 0x9, 0x0, 0x0)
	/usr/local/go/src/io/io.go:307 +0xa9
io.ReadFull(0x154b800, 0xc42d2d90e0, 0xc422910770, 0x5, 0x5, 0x46f2bd, 0x0, 0xc430257558)
	/usr/local/go/src/io/io.go:325 +0x58
github.com/weaveworks/cortex/vendor/google.golang.org/grpc.(*parser).recvMsg(0xc422910760, 0x7fffffff, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc430257498)
	/go/src/github.com/weaveworks/cortex/vendor/google.golang.org/grpc/rpc_util.go:218 +0x69
github.com/weaveworks/cortex/vendor/google.golang.org/grpc.recv(0xc422910760, 0x1557d40, 0x15b30b0, 0xc42d2d90e0, 0x0, 0x0, 0xd6b2a0, 0xc4229107a0, 0x7fffffff, 0x0, ...)
	/go/src/github.com/weaveworks/cortex/vendor/google.golang.org/grpc/rpc_util.go:314 +0x4d
github.com/weaveworks/cortex/vendor/google.golang.org/grpc.(*clientStream).RecvMsg(0xc4254929a0, 0xd6b2a0, 0xc4229107a0, 0x0, 0x0)
	/go/src/github.com/weaveworks/cortex/vendor/google.golang.org/grpc/stream.go:376 +0xe6
github.com/weaveworks/cortex/pkg/chunk/gcp.(*instrumentedClientStream).RecvMsg(0xc425c4d5c0, 0xd6b2a0, 0xc4229107a0, 0xc42acc3ba0, 0x155a9e0)
	/go/src/github.com/weaveworks/cortex/pkg/chunk/gcp/instrumentation.go:91 +0x63
github.com/weaveworks/cortex/vendor/google.golang.org/genproto/googleapis/bigtable/v2.(*bigtableMutateRowsClient).Recv(0xc42acc3ba0, 0x7f2fd915e178, 0xc42690bcb0, 0xc42690bf80)
	/go/src/github.com/weaveworks/cortex/vendor/google.golang.org/genproto/googleapis/bigtable/v2/bigtable.pb.go:879 +0x62
github.com/weaveworks/cortex/vendor/cloud.google.com/go/bigtable.(*Table).doApplyBulk(0xc422910500, 0x7f2fd915e178, 0xc42690bcb0, 0xc421923620, 0xc, 0xc, 0x0, 0x0, 0x0, 0x553499, ...)
	/go/src/github.com/weaveworks/cortex/vendor/cloud.google.com/go/bigtable/bigtable.go:671 +0x299
github.com/weaveworks/cortex/vendor/cloud.google.com/go/bigtable.(*Table).ApplyBulk.func1(0x7f2fd915e178, 0xc42690bcb0, 0x0, 0x0)
	/go/src/github.com/weaveworks/cortex/vendor/cloud.google.com/go/bigtable/bigtable.go:603 +0x88
github.com/weaveworks/cortex/vendor/cloud.google.com/go/bigtable/internal/gax.invokeWithRetry(0x7f2fd915e178, 0xc42690bcb0, 0xc42690bf20, 0x0, 0xc42690bf50, 0x5f5e100, 0x77359400, 0x3ff3333333333333, 0x0, 0x0, ...)
	/go/src/github.com/weaveworks/cortex/vendor/cloud.google.com/go/bigtable/internal/gax/invoke.go:53 +0x100
github.com/weaveworks/cortex/vendor/cloud.google.com/go/bigtable/internal/gax.Invoke(0x7f2fd915e178, 0xc42690bcb0, 0xc42690bf20, 0x158fa00, 0x2, 0x2, 0xc, 0xc)
	/go/src/github.com/weaveworks/cortex/vendor/cloud.google.com/go/bigtable/internal/gax/invoke.go:81 +0xc8
github.com/weaveworks/cortex/vendor/cloud.google.com/go/bigtable.(*Table).ApplyBulk(0xc422910500, 0x7f2fd915e178, 0xc42690bcb0, 0xc428353d40, 0xc, 0xc, 0xc421923560, 0xc, 0xc, 0x0, ...)
	/go/src/github.com/weaveworks/cortex/vendor/cloud.google.com/go/bigtable/bigtable.go:615 +0x541
github.com/weaveworks/cortex/pkg/chunk/gcp.(*storageClient).BatchWrite(0xc4201d4400, 0x7f2fd915e178, 0xc427c28120, 0x154d040, 0xc42690b8c0, 0x1, 0x154d040)
	/go/src/github.com/weaveworks/cortex/pkg/chunk/gcp/storage_client.go:96 +0x511
github.com/weaveworks/cortex/pkg/chunk.(*Store).updateIndex(0xc4201c9280, 0x7f2fd915e178, 0xc427c28120, 0xc420744140, 0xa, 0xc421962000, 0x1, 0x1, 0x2, 0x155c680)
	/go/src/github.com/weaveworks/cortex/pkg/chunk/chunk_store.go:113 +0xd5
github.com/weaveworks/cortex/pkg/chunk.(*Store).Put(0xc4201c9280, 0x7f2fd915e178, 0xc427c28120, 0xc421962000, 0x1, 0x1, 0x15f5aea5bb7, 0x15f5c49ebb7)
	/go/src/github.com/weaveworks/cortex/pkg/chunk/chunk_store.go:104 +0x18d
github.com/weaveworks/cortex/pkg/ingester.(*Ingester).flushChunks(0xc420474000, 0x7f2fd915e178, 0xc427c28120, 0x5d7f7098cc991a81, 0xc42f7463c0, 0xc423c77b60, 0x1, 0x2, 0xc430257f30, 0x154fbc0)
	/go/src/github.com/weaveworks/cortex/pkg/ingester/ingester_flush.go:197 +0x2fd
github.com/weaveworks/cortex/pkg/ingester.(*Ingester).flushUserSeries(0xc420474000, 0xc420744140, 0xa, 0x5d7f7098cc991a81, 0x0, 0x0, 0x0)
	/go/src/github.com/weaveworks/cortex/pkg/ingester/ingester_flush.go:170 +0x21f
github.com/weaveworks/cortex/pkg/ingester.(*Ingester).flushLoop(0xc420474000, 0x15)
	/go/src/github.com/weaveworks/cortex/pkg/ingester/ingester_flush.go:120 +0xe4
created by github.com/weaveworks/cortex/pkg/ingester.New
	/go/src/github.com/weaveworks/cortex/pkg/ingester/ingester.go:273 +0xce2
```
- there was nothing of note in the logs.
- the exact code this is running: https://github.com/tomwilkie/cortex/tree/ed6fcc89a9988917b4a43a05a7eeae1ba4a4059f
- upstream cortex issue: https://github.com/weaveworks/cortex/issues/586
- its gRPC v1.3.0 and v0.8.0 of this library

The context passed to ApplyBulk had no timeout applied, so I've added one (https://github.com/weaveworks/cortex/pull/587) but its my understanding this shouldn't hang indefinitely.

**Comment by garye** [comment]
Some of this has been mentioned in the cortex issue, but this doesn't necessarily seem like an actual deadlock. The Cloud Bigtable server has a default deadline on all requests so, from its perspective, only a previously undiscovered bug would prevent it from sending a response after a little while. It's possible that the response was dropped in some way in between Cloud Bigtable and the client library.

We'll try to see if we can construct a reproducible case. Any help you can provide with that would be much appreciated! Setting a deadline on requests like you added make a lot of sense and converts any future occurrence of this bug from a hang into an error that can be retried or at least logged.

**Comment by mediocregopher** [comment]
Hi, we're seeing this same issue pretty consistently (it's happened multiple times over the last few days):

```
goroutine 2838 [chan receive, 681 minutes]:
google.golang.org/grpc/transport.(*Stream).RecvCompress(...)
        /opt/TeamCityBuild/work/rachel-3/8232227c0afde646/.deps/src/google.golang.org/grpc/transport/transport.go:255
google.golang.org/grpc.(*clientStream).RecvMsg(0xc42cb36240, 0xbfe2e0, 0xc46f8e65e0, 0x0, 0x0)
        /opt/TeamCityBuild/work/rachel-3/8232227c0afde646/.deps/src/google.golang.org/grpc/stream.go:428 +0xa4f
google.golang.org/genproto/googleapis/bigtable/v2.(*bigtableMutateRowsClient).Recv(0xc44ffd0c80, 0x1173e00, 0xc4421e2ae0, 0xc46f8e5830)
        /opt/TeamCityBuild/work/rachel-3/8232227c0afde646/.deps/src/google.golang.org/genproto/googleapis/bigtable/v2/bigtable.pb.go:878 +0x62
cloud.google.com/go/bigtable.(*Table).doApplyBulk(0xc46d0a46c0, 0x1173e00, 0xc4421e2ae0, 0xc457cb7000, 0x2ec, 0x2ec, 0x0, 0x0, 0x0, 0x4d9a1c, ...)
        /opt/TeamCityBuild/work/rachel-3/8232227c0afde646/.deps/src/cloud.google.com/go/bigtable/bigtable.go:681 +0x43a
cloud.google.com/go/bigtable.(*Table).ApplyBulk.func1(0x1173e00, 0xc4421e2ae0, 0x0, 0x1)
        /opt/TeamCityBuild/work/rachel-3/8232227c0afde646/.deps/src/cloud.google.com/go/bigtable/bigtable.go:613 +0x88
cloud.google.com/go/bigtable/internal/gax.invokeWithRetry(0x1173e00, 0xc4421e2ae0, 0xc46f8e57d0, 0x0, 0xc46f8e5800, 0x5f5e100, 0x77359400, 0x3ff3333333333333, 0x0, 0x0, ...)
        /opt/TeamCityBuild/work/rachel-3/8232227c0afde646/.deps/src/cloud.google.com/go/bigtable/internal/gax/invoke.go:53 +0x100
cloud.google.com/go/bigtable/internal/gax.Invoke(0x1173e00, 0xc4421e2ae0, 0xc46f8e57d0, 0x11d5300, 0x2, 0x2, 0x2ec, 0x2ec)
        /opt/TeamCityBuild/work/rachel-3/8232227c0afde646/.deps/src/cloud.google.com/go/bigtable/internal/gax/invoke.go:81 +0xc8
cloud.google.com/go/bigtable.(*Table).ApplyBulk(0xc46d0a46c0, 0x1173e00, 0xc4421e2ae0, 0xc422bf7000, 0x2ec, 0x2ec, 0xc457cb4000, 0x2ec, 0x2ec, 0x0, ...)
        /opt/TeamCityBuild/work/rachel-3/8232227c0afde646/.deps/src/cloud.google.com/go/bigtable/bigtable.go:612 +0x511
bitbucket.org/levenlabs/analytics-api/analyticsapi/aasub.(*BigTableDB).upsertByTableDirect.func2()
        /opt/TeamCityBuild/work/rachel-3/8232227c0afde646/.deps/src/bitbucket.org/levenlabs/analytics-api/analyticsapi/aasub/db_bt.go:298 +0x11e
bitbucket.org/levenlabs/llib/workgroup.(*WorkGroup).spin(0xc4230d2f10)
        /opt/TeamCityBuild/work/rachel-3/8232227c0afde646/.deps/src/bitbucket.org/levenlabs/llib/workgroup/workgroup.go:111 +0x45
created by bitbucket.org/levenlabs/llib/workgroup.New
        /opt/TeamCityBuild/work/rachel-3/8232227c0afde646/.deps/src/bitbucket.org/levenlabs/llib/workgroup/workgroup.go:102 +0xd7
```

```
cloud.google.com/go (git) - 442eb1f
github.com/googleapis/gax-go (git) - 317e000
google.golang.org/grpc (git) - cd563b8
```

This is something which started pretty recently, we hadn't seen it up until last week or so. The context we're passing to ApplyBulk does have a timeout, however we're using the same context for multiple serial ApplyBulk calls. If the context was to timeout on the first ApplyBulk, the following calls would get a canceled context and might not handle that correctly. That's my current theory, I'm going to update our code to check for the context being cancelled between ApplyBulks and see if that has an effect.

**Comment by garye** [comment]
That's interesting. I'm not sure there is anything specific that the bigtable library should be doing in the case of a cancelled context. @jba Any thoughts on that?

**Comment by jba** [comment]
A gRPC call with a cancelled context should return immediately with an error. Unless the bigtable client is doing some long-running activity, it shouldn't need to check for cancelled contexts itself.

**Comment by garye** [comment]
Thanks @jba. @mediocregopher, it might be helpful to open an issue with gRPC about this to see if they have any ideas.

**Comment by mediocregopher** [comment]
Update: the context thing might have been a red-herring. We added the following before each ApplyBulk call:

```
		if deadline, ok := ctx.Deadline(); ok && deadline.Sub(time.Now()) < time.Second {
			return errors.New("context doesn't have long to live, bailing")
		}
```

so unless ApplyBulk is doing something which takes more than a second before the grpc call the context shouldn't be cancelled. But this weekend we saw two more instances of ApplyBulk hanging on RecvCompress as in that stacktrace.

I'll post an issue in the gRPC project as well though.
